### ISSUE TITLE
[MODULES-153] Ability to discover all jars available to a generic application for a non-running server with a given config file

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#505]
@robstryker opened **[MODULES-153](https://issues.redhat.com/browse/MODULES-153?redirect=false)** and commented

Not sure if this is 100% the domain of modules, or if it bleeds into AS. Either way:

Tooling is looking for a stable way to get a list of jars accessible for a generic project (web, sar, custom) which will be deployed to the AS, for the purposes of modifying an eclipse project's classpath to provide these jars in the workspace.

The current method under investigation is a PoC found here:  https://github.com/rafabene/jboss-modules-poc, though we intend to bend it to our own purposes of course ;)

We have been advised that simply poking around the modules folder is not a sustainable or guaranteed way to acquire this list of jars, and so we're looking for something a bit more stable.

The idea here would be to generate a classpath that would assist the developer in developing his application. We are looking to avoid jars with duplicate class entries (jsf1, jsf2, etc etc).

It would be best if we could somehow limit this to only what jars the application will be able to access, but I am not sure if this is feasible considering the application is of indeterminate type (sar or web, etc, we do not expect jboss-modules to be able to figure that out or even respond should we provide it. My feeling is that would not be feasible for jboss-modules)

So again, we'd like some solution that is not necessarily a minimal set of jars required, but at the very least contains no duplicates that would hinder the developer's coding progress. (Duplicate entries in eclipse could cause errors on deployment if, for example, it was compiled against jsf1 but was deployed to jsf2)

In general we do not intend to bundle jboss-modules.jar at all since it is likely to change substantially from version to version as new features are added, so tools will most likely access this via reflection (loading the jar via a classloader in an isolated thread and then disposing of ti later after our query has finished).

For this reason we would like at least this jira's suggested api to be stable and backwards compatible.

So, basically, given a jboss installtion (or modules folder), and a configuration file (standalone.xml or similar), we would like a list of jars we could add to a projects classpath.

Is this do-able? :)

---

**Issue Links:**
- [JBIDE-13139](https://issues.redhat.com/browse/JBIDE-13139) Improve libs import for EAP 6
 (_**"blocks"**_)
- [WFLY-1026](https://issues.redhat.com/browse/WFLY-1026) Require method to determine best-guess classpath for projects targeted to non-running server

-1 votes, 5 watchers

**Comment by dmlloyd** [comment]
@bstansberry commented

An aspect that definitely bleeds into AS is the configuration file part. JBoss Modules has no idea about AS configuration files.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Generally speaking, no, this is not possible.  However, the list of modules available to a deployment is really specific to the policy of AS.  Modules just carries out the linking that it is given; AS is what makes the decisions as to what gets connected where.  This decision is affected primarily by what subsystems are installed, which Modules has no control or knowledge of.

I think it would be (at least conceptually) possible to create an AS operation which assembles and returns this information without actually executing the deployment, but it would not be trivial.

**Comment by dmlloyd** [comment]
@robstryker commented

Well again, one of our requirements is to not have any running server, and a minimal set of jars required to run the task. But we also don't need 100% strict classpath matching to what it would be on a running server for that specific module. A superset is acceptable so long as theres no possibility of conflicts.

Basically, we just want a best-guess that doesn't leave out any obviously available jars and avoids duplicates that would lead to nondeterministic or incorrect compilations.

Currently we hard-code the classpath by hand-picking a dozen or so jars that we feel are API (or required impls) and all projects would make use of. But this clearly is not configurable enough for a user, and if a user adds a module to their standalone.xml, we have no way of ensuring that it's automatically added to an eclipse project's classpath.

We're not looking for perfection. We're looking for better than what we have currently.

**Comment by dmlloyd** [comment]
@dmlloyd commented

JBoss Modules by itself has less information than you have.  I think this would have to be solved in AS, if anywhere (though I doubt that we can solve it there either).

**Comment by dmlloyd** [comment]
@bstansberry commented

For sure the entire problem can't be solved in JBoss Modules. Part of it needs to be in the tools or the AS. And maybe none of it should be in the AS, since the fairly loose set of requirements for what is acceptable as the "classpath" sounds like a feature that should be constrained to the tools and not something exposed as an API by the AS. Once it becomes an AS API it becomes open for misuse by completely unforeseen people.

The part that I don't think we want either tools or the AS doing is poking around in the modules/ dir. It's bad to do that now, and once AS7-6344 is in place it's worse. And then when patching is in place, it's even worse. JBoss Modules will be the layer that has the understanding of what locations in the module repo will actually be relevant in the running server.

In a coarse grained way, I see this as:

1) Tools or AS decides what modules are relevant.
2) Tools or AS asks JBoss Modules for information about those modules.
3) JBoss Modules finds the modules and provides info via a stable API.
4) Tools or AS uses info to decide whether the given module is relevant (e.g. looks at the public/private flag) and to find the URLs of the relevant JARs.

So the design of the overall feature should be to decide what 1-4 are all about, and then this particular JIRA becomes the details for 3).

**Comment by dmlloyd** [comment]
**[Max Andersen](https://issues.redhat.com/secure/ViewProfile.jspa?name=maxandersen)** commented

My 2 cent on this:

In osgi there are specs/rules for how the resolution is done based on metadata in jar + the list of available other bundles - that is something eclipse already have support for. I was hoping JBoss Modules+AS could have something similar. Preferably without having to actually launch the server - but we could work around that and just start up the server once to get some data to cache for project creation etc.

That said I think it would be superuseful that AS/Modules had a way to report back which jars/modules/packages are **actually** activated for a certain deployment. **Especially** when AS7-6344 comes in play with multiple locations/layers.

And yes, I think the approach Brian outlined could work.
Note: the module is always relevant, public/private I expect to just use for setting access restrictions so users gets warnings (or errors if they decide) on these.

For #1 what we got in Tools we got a bunch of metadata to use to calculate the right classpath. We got a list of so called facets that the user is targeting/adding. i.e. JPA 2.0, Web 3.0, Java 7.0 etc. that I could imagine us making a mapping from to set of modules.

That could at least be used for a "better educated guess".

To be "more correct" we would also need to provide the actual manifest.mf which jboss modules looks at to add/remove certain modules.

**Comment by dmlloyd** [comment]
**[Max Andersen](https://issues.redhat.com/secure/ViewProfile.jspa?name=maxandersen)** commented

To clarify - the workflow for tools is that you can use the runtime before actually starting the server and that goes for all other tools/runtimes I know afaik so would be great if this could all be done - at least for the best guess without launching a full server.

**Comment by dmlloyd** [comment]
@robstryker commented

So here's my idea.

Would it be possible if I give jboss-modules in a non-running server  1) a modules directory, and 2) a module string (such as org.jboss.as.web), it would return me a basic model or simply a list of best-guess jars? We would be able to provide overrides, for example  "com.sun.jsf-impl, slot 1.2"

I could read / parse the standalone.xml myself to get the list of top-level modules we want. I forsee something like this, but this is just an example.

```
   String[] modulesINeed = new String[]{"org.jboss.as.web" };
   File[] allJars = jbossModulesGetJarRecurse(modulesINeed);

OR
     String[] modulesINeed = new String[]{"org.jboss.as.web" };
     String[] overrideId = new String[] {"com.sun.jsf-impl" };
     String[] overrideSlot = new String[] { "1.2" };
     File[] allJars = jbossModulesGetJarRecurse(modulesINeed, overrideId, overrideSlot);
```

We could also have overrideSlot be some type of osgi version string with ranges, if possible.

As a return value, we would expect a list of all dependencies (recursive included) in a best-guess manner. Instead of File[], we could return some object array which would include the File, and information on its restrictions, whether the jar itself is public / private, or if public / private is enhanced to be per-package instead of per jar, some list of restricted packages.

In this way, it would be up to tools to get the list of top module names, either by asking the user directly in our UI or parsing the standalone.xml ourselves. jboss-modules job is a lot easier, a lot less busy, AS itself requires no additions to api, and tools still maintains some responsibility at the top level. And best of all, a started server is not required.

Does this sound feasible? What work would need to be done? Is it possible I can help? ;)

(This message edited for formatting and additional info)

**Comment by dmlloyd** [comment]
@robstryker commented

```
In a coarse grained way, I see this as:

1) Tools or AS decides what modules are relevant.
2) Tools or AS asks JBoss Modules for information about those modules.
3) JBoss Modules finds the modules and provides info via a stable API.
4) Tools or AS uses info to decide whether the given module is relevant (e.g. looks at the public/private flag) and to find the URLs of the relevant JARs.
```

This sounds fine to me in general. We're starting to push up against time-constraints though, so I think we need to start taking this issue a bit more seriously. Any chance we can flesh out an API for this?

**Comment by dmlloyd** [comment]
@dmlloyd commented

I believe that our dependency metadata and/or management information contains all of the information needed to solve this outside of the core of JBoss Modules. Additionally I think the demand for this feature has also dried up.
